vnext/ReactWindowsCore/WinRTWebSocketResource.h

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+
+#pragma once
+
+#include <winrt/Windows.Networking.Sockets.h>
+#include <winrt/Windows.Storage.Streams.h>
+#include <IWebSocketResource.h>
+
+#include <future>
+#include <queue>
+
+namespace Microsoft::React
+{
+
+class WinRTWebSocketResource : public IWebSocketResource
+{
+  winrt::Windows::Foundation::Uri m_uri;
+  winrt::Windows::Networking::Sockets::MessageWebSocket m_socket;
+  winrt::Windows::Networking::Sockets::MessageWebSocket::MessageReceived_revoker m_revoker;
+  winrt::Windows::Storage::Streams::DataWriter m_writer{ m_socket.OutputStream() };
+
+  std::atomic_bool m_connectRequested;
+  std::atomic_bool m_sendRequested;
+  std::promise<void> m_connectPerformed;
+  std::promise<void> m_closePerformed;
+  winrt::Windows::Foundation::IAsyncAction m_connectAction;
+
+  CloseCode m_closeCode{ CloseCode::None };
+  std::string m_closeReason;
+  std::queue<std::pair<std::string, bool>> m_writeQueue;
```

Comment:
As it is, this will have poor cache locality and be wasteful in terms of memory usage. The bool will take up at least 4 bytes for every entry. If there will be a lot of entries in the queue, we should optimize this further and create a struct to keep track of the stuff in this queue. (vector<bool> is optimized for doing bit twiddling so you end up using the full 8 bits/byte, you could leverage that)